src/psyclone/tests/nemo/transformations/openmp/openmp_test.py

```diff
@@ -152,8 +152,7 @@ def test_omp_parallel_multi():
     # gives elements 2-3).
     otrans.apply(schedule[0].loop_body[2:4])
     gen_code = str(psy.gen).lower()
-    assert ("    !$omp parallel default(shared), private(ji,jj,zabe1,zcof1,"
```

Comment:
I am happy to rethink this, but note that this is not new. In the _get_private_clause there was already this comment:
```
# Check if the write access is inside the parallel loop. If
# the write is outside of a loop, it is an assignment to
# a shared variable. Example where jpk is likely used
# outside of the parallel section later, so it must be
# declared as shared in order to have its value in other loops:
# !$omp parallel
# jpk = 100
# !omp do
# do ji = 1, jpk

# TODO #598: improve the handling of scalar variables.
```
and this is also hinted as several places, e.g: src/psyclone/tests/nemo/tests_files/explicit_do_two_loops.f90
and the #598 starting comment.

Reply:
Maybe @hiker can provide some context? Was this the behavior at some point or just marking a potential issue?

Reply:
I suppose this is a bit of a weird test case - we're putting a (nested) loop inside a `parallel` region but are not parallelising the loop. Previously we were at least saying that the loop variables were private (which seems sensible). I suppose the question is, why are we now saying that the loop variables are shared and should we change this?

Reply:
I've checked the OMP spec. and I think I'm right in saying that what we currently generate in this example will mean that the loop variables are shared between threads.

Reply:
I would have thought the opposite:
Section 5.1:
> Loop iteration variables inside parallel, teams, or task generating constructs are private in the innermost such construct that encloses the loop.

That would need a bit of verification/a lawyer to read the details. A similar statement is above for C/C++ (afaik Fortran kind of requires any loop variable to be private, since the number of iterations is pre-determined at the start of the loop - as in a non-parallelised loop in an inner region).

Having said that, I would actually prefer to have all loop variables listed as private, since this avoid the head scratching if this is a bug or not :)

Reply:
So you are kind of both right. I have been playing with it and the loop variable indeed defaults to private and is not a problem in the previous example. But "would be private" variables inside the loop are shared with the current implementation, so in:
```
omp parallel
scalar1 = 1
do i= start, stop
    scalar2 = <compute>
    array(i) = scalar2
enddo
omp end do
```
there is a race condition on scalar2 and the results will end up different than in the serial semantics. So I propose to keep top-level scalar reference as shared (as all threads write the same) and loop/ifs bodies scalar assignments as private (I can add a transformationError if they are used anywhere after the parallel region)

@hiker, @arporter What do you think?

Reply:
That sounds sensible to me. When you say 'if bodies', presumably you only mean within loops? Also, this doesn't have implications for our ability to identify reductions does it?

Reply:
(And thanks @hiker for managing to read the spec more carefully than I did!)